src/ring_theory/witt_vector/compare.lean

```diff
@@ -0,0 +1,199 @@
+/-
+Copyright (c) 2020 Johan Commelin. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Johan Commelin, Robert Y. Lewis
+-/
+
+import ring_theory.witt_vector.truncated
+import ring_theory.witt_vector.identities
+import data.padics.ring_homs
+
+/-!
+
+# Comparison isomorphism between `witt_vector p (zmod p)` and `ℤ_[p]`
+
+We construct a ring isomorphism between `witt_vector p (zmod p)` and `ℤ_[p]`.
+This isomorphism follows from the fact that both satisfy the universal property
+of the inverse limit of `zmod (p^n)`.
+
+## Main declarations
+
+* `witt_vector.to_zmod_pow`: a family of compatible ring homs `𝕎 (zmod p) → zmod (p^k)`
+* `witt_vector.equiv`: the isomorphism
+
+-/
+
+noncomputable theory
+
+namespace truncated_witt_vector
+
+variables {p : ℕ} [hp : fact p.prime] (n : ℕ) (R : Type*) [comm_ring R]
+include hp
+
+local notation `𝕎` := witt_vector p -- type as `\bbW`
+
+variables (p n R)
+
+lemma eq_of_le_of_cast_pow_eq_zero [char_p R p] (i : ℕ) (hin : i ≤ n)
+  (hpi : (p ^ i : truncated_witt_vector p n R) = 0) :
+  i = n :=
+begin
+  contrapose! hpi,
+  replace hin := lt_of_le_of_ne hin hpi, clear hpi,
+  have : (↑p ^ i : truncated_witt_vector p n R) = witt_vector.truncate n (↑p ^ i),
+  { rw [ring_hom.map_pow, ring_hom.map_nat_cast] },
+  rw [this, ext_iff, not_forall], clear this,
+  use ⟨i, hin⟩,
+  rw [witt_vector.coeff_truncate, coeff_zero, fin.coe_mk, witt_vector.coeff_p_pow],
+  haveI : nontrivial R := char_p.nontrivial_of_char_ne_one hp.ne_one,
+  exact one_ne_zero
+end
+
+section iso
+
+variables (p n) {R}
+
+lemma card_zmod : fintype.card (truncated_witt_vector p n (zmod p)) = p ^ n :=
+by rw [card, zmod.card]
+
+lemma char_p_zmod : char_p (truncated_witt_vector p n (zmod p)) (p ^ n) :=
+char_p_of_prime_pow_injective _ _ _ (card_zmod _ _)
+    (eq_of_le_of_cast_pow_eq_zero p n (zmod p))
+
+local attribute [instance] char_p_zmod
+
+/--
+Since `truncated_witt_vector p n (zmod p)` is a finite ring
+with characteristic and cardinality `p^n`,
+it is isomorphic to `zmod (p^n)`.
+-/
+def zmod_equiv_trunc : zmod (p^n) ≃+* truncated_witt_vector p n (zmod p) :=
+zmod.ring_equiv (truncated_witt_vector p n (zmod p)) (card_zmod _ _)
+
+lemma zmod_equiv_trunc_apply {x : zmod (p^n)} :
+  zmod_equiv_trunc p n x =
+  zmod.cast_hom (by refl) (truncated_witt_vector p n (zmod p)) x :=
+rfl
+
+lemma commutes {m : ℕ} (hm : n ≤ m) :
+  (truncate hm).comp (zmod_equiv_trunc p m).to_ring_hom =
+    (zmod_equiv_trunc p n).to_ring_hom.comp (zmod.cast_hom (pow_dvd_pow p hm) _) :=
```

Comment:
Done. The proof can not be inlined, because it does definitional unfolding, but the lemma is used by `rw`. So this lemma is providing a useful syntactic variant.